README.md
# Wattpilot

> :warning: This project is still in early development and might never leave this state

`wattpilot` is a Python 3 (>= 3.10) module to interact with Fronius Wattpilot wallboxes which do not support (at the time of writting) a documented API. This functionality of this module utilized a undocumented websockets API, which is also utilized by the official Wattpilot.Solar mobile app.

## Wattpilot API Documentation

See [API.md](API.md) for the current state of the API documentation this implementation is based on.

It has been compiled from different sources, but primarily from:

* [go-eCharger-API-v1](https://github.com/goecharger/go-eCharger-API-v1/blob/master/go-eCharger%20API%20v1%20EN.md)
* [go-eCharger-API-v2](https://github.com/goecharger/go-eCharger-API-v2/blob/main/apikeys-en.md)

## Wattpilot Shell

The shell provides an easy way to explore the available properties and get or set their values.

```bash
# Install the wattpilot module, if not yet done so:
pip install .
```

Run the interactive shell

```bash
# Usage:
export WATTPILOT_HOST=<wattpilot_ip>
export WATTPILOT_PASSWORD=<password>
wattpilotshell
Welcome to the Wattpilot Shell 0.2.   Type help or ? to list commands.

wattpilot> help

Documented commands (type help <topic>):
========================================
EOF      exit  ha    info  properties  server  unwatch  watch
connect  get   help  mqtt  rawvalues   set     values 
```

The shell supports TAB-completion for all commands and their arguments.
Detailed documentation can be found in [ShellCommands.md](ShellCommands.md).

It's also possible to pass a single command to the shell to integrate it into scripts:

```bash
# Usage:
wattpilotshell "<command> <args...>"

# Examples:
wattpilotshell "get amp"
wattpilotshell "set amp 6"
```

## MQTT Bridge Support

It is possible to publish JSON messages received from Wattpilot and/or individual property value changes to an MQTT server.
The easiest way to start the shell with MQTT support is using these environment variables:

```bash
export MQTT_ENABLED=true
export MQTT_HOST=<mqtt_host>
export WATTPILOT_HOST=<wattpilot_ip>
export WATTPILOT_PASSWORD=<wattpilot_password>
wattpilotshell
```

Pay attention to environment variables starting with `MQTT_` to fine-tune the MQTT support (e.g. which messages or properties should published to MQTT topics).

MQTT support can be easily tested using mosquitto:

```bash
# Start mosquitto in a separate console:
mosquitto

# Subscribe to topics in a separate console:
mosquitto_sub -t 'wattpilot/#' -v
```

## Home Assistant MQTT Discovery Support

To enable Home Assistant integration (using MQTT) set `MQTT_ENABLED` and `HA_ENABLED` to `true` and make sure to correctly configure the [MQTT Integration](https://www.home-assistant.io/integrations/mqtt).
It provides auto-discovery of entities using property configuration from [wattpilot.yaml](src/wattpilot/ressources/wattpilot.yaml).
The is the simplest possible way to start the shell with HA support:

```bash
export MQTT_ENABLED=true
export HA_ENABLED=true
export MQTT_HOST=<mqtt_host>
export WATTPILOT_HOST=<wattpilot_ip>
export WATTPILOT_PASSWORD=<wattpilot_password>
wattpilotshell
```

Pay attention to environment variables starting with `HA_` to fine-tune the Home Assistant integration (e.g. which properties should be exposed).

The discovery config published to MQTT can be tested using this in addition to the testing steps from MQTT above:

MQTT support can be easily tested using mosquitto:

```bash
# Subscribe to homeassisant topics in a separate console:
mosquitto_sub -t 'homeassistant/#' -v
```

## Docker Support

The Docker images for the Wattpilot MQTT bridge with Home Assistant MQTT discovery can be found on [GitHub Packages](https://github.com/joscha82/wattpilot/pkgs/container/wattpilot):

```bash
# Pull Image:
docker pull ghcr.io/joscha82/wattpilot:latest
# NOTE: Use the tag 'latest' for the latest release, a specific release version or 'main' for the current image of the not yet released main branch.

# Create .env file with environment variables:
cat .env
HA_ENABLED=true
MQTT_ENABLED=true
MQTT_HOST=<mqtt_host>
WATTPILOT_HOST=<wattpilot_ip>
WATTPILOT_PASSWORD=<my_secret_password>

# Run container (recommended with MQTT_ENABLED=true and HA_ENABLED=true - e.g. on a Raspberry Pi):
docker-compose up -d
```

To diagnose the hundreds of Wattpilot parameters the shell can be started this way (typically recommended with `MQTT_ENABLED=false` and `HA_ENABLED=false` on a local machine, in case a Docker container with MQTT support may be running permanently on e.g. a Raspberry Pi):

```bash
# Create .env file with environment variables:
cat .env
HA_ENABLED=false
MQTT_ENABLED=false
MQTT_HOST=<mqtt_host>
WATTPILOT_HOST=<wattpilot_ip>
WATTPILOT_PASSWORD=<my_secret_password>

# Run the shell:
docker-compose run wattpilot shell
```

## Environment Variables

| Environment Variable        | Description                                                                                                                                                                                  | Default Value                                 |
| --------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------- |
| `HA_ENABLED`                | Enable Home Assistant Discovery                                                                                                                                                              | `false`                                       |
| `HA_PROPERTIES`             | Only discover given properties (leave unset for all properties having `homeAssistant` set in [wattpilot.yaml](src/wattpilot/ressources/wattpilot.yaml))                                                                               |                                               |
| `HA_TOPIC_CONFIG`           | Topic pattern for HA discovery config                                                                                                                                                        | `homeassistant/{component}/{uniqueId}/config` |
| `HA_WAIT_INIT_S`            | Wait initial number of seconds after starting discovery (in addition to wait time depending on the number of properties). May be increased, if entities in HA are not populated with values. | `5`                                           |
| `HA_WAIT_PROPS_MS`          | Wait milliseconds per property after discovery before publishing property values. May be increased, if entities in HA are not populated with values.                                         | `50`                                          |
| `MQTT_AVAILABLE_PAYLOAD`    | Payload for the availability topic in case the MQTT bridge is online                                                                                                                                                                                | `online`                              |
| `MQTT_CLIENT_ID`            | MQTT client ID                                                                                                                                                                               | `wattpilot2mqtt`                              |
| `MQTT_ENABLED`              | Enable MQTT                                                                                                                                                                                  | `false`                                       |
| `MQTT_HOST`                 | MQTT host to connect to                                                                                                                                                                      |                                               |
| `MQTT_MESSAGES`             | List of space-separated message types to be published to MQTT (leave unset for all messages)                                                                                                 |                                               |
| `MQTT_NOT_AVAILABLE_PAYLOAD` | Payload for the availability topic in case the MQTT bridge is offline (last will message)                                                                                                                                                                               | `offline`                              |
| `MQTT_PORT`                 | Port of the MQTT host to connect to                                                                                                                                                          | `1883`                                        |
| `MQTT_PROPERTIES`           | List of space-separated property names to publish changes for (leave unset for all properties)                                                                                               |                                               |
| `MQTT_PUBLISH_MESSAGES`     | Publish received Wattpilot messages to MQTT                                                                                                                                                  | `false`                                       |
| `MQTT_PUBLISH_PROPERTIES`   | Publish received property values to MQTT                                                                                                                                                     | `true`                                        |
| `MQTT_TOPIC_AVAILABLE`      | Topic pattern to publish Wattpilot availability status to                                                                                                                                               | `{baseTopic}/available`          |
| `MQTT_TOPIC_BASE`           | Base topic for MQTT                                                                                                                                                                          | `wattpilot`                                   |
| `MQTT_TOPIC_MESSAGES`       | Topic pattern to publish Wattpilot messages to                                                                                                                                               | `{baseTopic}/messages/{messageType}`          |
| `MQTT_TOPIC_PROPERTY_BASE`  | Base topic for properties                                                                                                                                                                    | `{baseTopic}/properties/{propName}`           |
| `MQTT_TOPIC_PROPERTY_SET`   | Topic pattern to listen for property value changes for                                                                                                                                       | `~/set`                                       |
| `MQTT_TOPIC_PROPERTY_STATE` | Topic pattern to publish property values to                                                                                                                                                  | `~/state`                                     |
| `WATTPILOT_AUTOCONNECT`     | Automatically connect to Wattpilot on startup                                                                                                                                                | `true`                                        |
| `WATTPILOT_AUTO_RECONNECT`  | Automatically re-connect to Wattpilot on lost connections                                                                                                                                            | `true`                                        |
| `WATTPILOT_CONNECT_TIMEOUT` | Connect timeout for Wattpilot connection                                                                                                                                                     | `30`                                          |
| `WATTPILOT_HOST`            | IP address of the Wattpilot device to connect to                                                                                                                                             |                                               |
| `WATTPILOT_INIT_TIMEOUT`    | Wait timeout for property initialization                                                                                                                                                     | `30`                                          |
| `WATTPILOT_LOGLEVEL`     | Log level (CRITICAL,ERROR,WARNING,INFO,DEBUG)                                                                                                                                                             | `INFO`                                        |
| `WATTPILOT_PASSWORD`        | Password for connecting to the Wattpilot device                                                                                                                                              |                                               |
| `WATTPILOT_RECONNECT_INTERVAL` | Waiting time in seconds before a lost connection is re-connected                                                                                                                                   | `30`                                          |
| `WATTPILOT_SPLIT_PROPERTIES` | Whether compound properties (e.g. JSON arrays or objects) should be decomposed into separate properties                                                                                      | `true`                                        |

## HELP improving API definition in wattpilot.yaml

The MQTT and Home Assistant support heavily depends on the API definition in [wattpilot.yaml](src/wattpilot/ressources/wattpilot.yaml) which has been compiled from different sources and does not yet contain a full set of information for all relevant properties.
See [API.md](API.md) for a generated documentation of the available data.

If you want to help, please have a look at the properties defined in [wattpilot.yaml](src/wattpilot/ressources/wattpilot.yaml) and fill in the missing pieces (e.g. `title`, `description`, `rw`, `jsonType`, `childProps`, `homeAssistant`, `device_class`, `unit_of_measurement`, `enabled_by_default`) to properties you care about.
The file contains enough documentation and a lot of working examples to get you started.


changelog.txt
Version 0.2 (2022-05-26)
- New Featuer: mqtt Bridge (@ahochsteger)
- Bugfixes

Version 0.1 (2022-04-20)
- New Featuer: shell (@ahochsteger)
- New Feature: mqtt Bridge in shell (@ahochsteger)

Version 0.0.1 (2022-02-17)
- Initial testing release

test-shell.sh
#!/bin/bash

# This script helps during development to test certain aspects of the wattpilot shell.

if [ -f dev.env ]; then
    source dev.env
fi

export PYTHONPATH=src

WPCONFIG_FILE="src/wattpilot/ressources/wattpilot.yaml"

cmd="${1:-default}"
shift 1

function runShell() {
    python -m wattpilot.wattpilotshell "${@}"
}

function runShellOnly() {
    WATTPILOT_AUTOCONNECT=false MQTT_ENABLED=false HA_ENABLED=false runShell "${@}"
}

function runShellWithProps() {
    PROPS="${1:-}"
    echo "Enabled properties: ${PROPS}"
    MQTT_ENABLED=true HA_ENABLED=true MQTT_PROPERTIES="${PROPS}" HA_PROPERTIES="${PROPS}" runShell
}

function runShellWithAllProps() {
    PROP_FILTER="${1:-.*}"
    PROPS=$(gojq --yaml-input -r -c '.properties[] | (.key, .childProps?[]?.key?)' "${WPCONFIG_FILE}" | grep "^${PROP_FILTER}" | xargs echo)
    runShellWithProps "${PROPS}"
}


case "${cmd}" in
    default)
        runShell "${@}"
    ;;
    shell-only)
        runShellOnly "${@}"
    ;;
    server)
        WATTPILOT_AUTOCONNECT=true MQTT_ENABLED=true HA_ENABLED=true runShell "server"
    ;;
    save)
        logfile=work/status-$(date +"%Y-%m-%d_%H-%M-%S")-${1:-adhoc}.log
        mkdir -p work
        WATTPILOT_DEBUG_LEVEL=WARNING MQTT_ENABLED=false HA_ENABLED=false runShell "values" >>${logfile} 2>&1
        WATTPILOT_DEBUG_LEVEL=WARNING MQTT_ENABLED=false HA_ENABLED=false runShell "rawvalues" >>${logfile} 2>&1
        WATTPILOT_DEBUG_LEVEL=WARNING MQTT_ENABLED=false HA_ENABLED=false runShell "properties" >>${logfile} 2>&1
    ;;
    ha)
        runShellWithProps
    ;;
    ha-all)
        runShellWithAllProps
    ;;
    ha-test-props)
        runShellWithProps "alw loe nrg fhz spl3 acu ama cdi ffna fna"
    ;;
    ha-test-array)
        runShellWithProps "nrg"
    ;;
    ha-test-boolean)
        runShellWithProps "alw loe"
    ;;
    ha-test-float)
        runShellWithProps "fhz spl3"
    ;;
    ha-test-integer)
        runShellWithProps "acu ama"
    ;;
    ha-test-object)
        runShellWithProps "cdi"
    ;;
    ha-test-string)
        runShellWithProps "ffna fna"
    ;;
    update-docs)
        python gen-apidocs.py >API.md
        (
            echo "# Wattpilot Shell Commands"
            for cmd in $(
                runShellOnly "help" \
                | awk 'BEGIN {p=0} {if(p) print $0} /^==/ {p=1}' \
                | xargs -n 1 echo \
                | grep -E -v '^EOF$' \
                | sort \
            ); do
                echo ""
                echo "## ${cmd}"
                echo ""
                echo "\`\`\`bash"
                runShellOnly "help ${cmd}"
                echo "\`\`\`"
            done
        ) >ShellCommands.md
        (
            # NOTE: The file cannot yet fully replace the table in README.md since the description is missing
            echo "# Wattpilot Shell Environment Variables"
            echo ""
            echo "| Environment Variable | Default Value |"
            echo "|----------------------|---------------|"
            cat src/wattpilot/wattpilotshell.py \
            | awk 'BEGIN {p=0} /^ +/ {if(p) print $0} /^def / {p=0} /^def main_setup_env\(\):/ {p=1}' \
            | grep -E -v '\b(global|assert)\b' \
            | sed -re 's/#.*//g;s/\n//g' \
            | tr '\n' ' ' \
            | sed -re "s/os\\.environ\\.get\\(\\s*'([A-Z_]+)'\\s*,\\s*'([^']*)'\\s*\\)/\n| \`\1\` | \`\2\` |\n/g; s/\`\`//g" \
            | grep -E '^\|' \
            | sort
        ) >ShellEnvVariables.md
    ;;
    validate-yaml)
        echo -n "Properties with missing mandatory fields: "
        gojq --yaml-input -c '[.properties[],.childProps[]? | select(.key==null or .jsonType==null)]' <"${WPCONFIG_FILE}"
        echo -n "Child properties with missing jsonType: "
        gojq --yaml-input -c '[.properties[] | . as $ppd | .childProps[]? | . as $cpd | select(.jsonType==null and $ppd.itemType==null) | .key]' <"${WPCONFIG_FILE}"
    ;;
    *)
        echo "Unknown command: ${cmd}"
        exit 1
    ;;
esac


gen-apidocs.py
import json
import yaml

WATTPILOT_API_FILE = "src/wattpilot/ressources/wattpilot.yaml"

# Property code formatting:
def pc(prop,key,alt=''):
    return f"`{prop[key]}`" if key in prop and prop[key] != "" else alt

# Property key/alias formatting:
def pk(prop):
    return pc(prop,'key','-') + "<br>" + pc(prop,'alias','-')

# Property type formatting:
def pt(prop):
    return pc(prop,'jsonType','-') + "<br>" + pc(prop,'type','-')

# Property value formatting:
def pv(prop,key,alt=''):
    return prop[key] if key in prop else alt

# Property Home Assistant info formatting:
def ha(prop):
    return ':heavy_check_mark:' if 'homeAssistant' in prop else ':white_large_square:'

# Message and property example formatting:
def e(d):
    return f"`{json.dumps(d['example'])}`" if "example" in d else ""

with open(WATTPILOT_API_FILE, 'r') as stream:
    wpapidef = yaml.safe_load(stream)

s = "# Wattpilot API Description\n\n"

s += "## WebSocket Message Types\n\n"
s += "| Key | Title | Description | Example |\n"
s += "|-----|-------|-------------|---------|\n"
for msg in wpapidef["messages"]:
    s += f"| `{msg['key']}` | {msg['title']} | {msg['description']} | {e(msg)} |\n"
s += "\n"

s += "## WebSocket API Properties\n\n"
s += "| Key/Alias | Title | R/W | JSON/API Type | Category | HA Enabled | Description | Example |\n"
s += "|-----------|-------|-----|---------------|----------|------------|-------------|---------|\n"
for p in wpapidef["properties"]:
    s += f"| {pk(p)} | {pv(p,'title')} | {pv(p,'rw')} | {pt(p)} | {pv(p,'category')} | {ha(p)} | {pv(p,'description')} | {e(p)} |\n"

print(s, end='')


shell.py
import argparse
import json
import logging
import os
import paho.mqtt.client as mqtt
import readline
import wattpilot
import yaml

from time import sleep
from types import SimpleNamespace

_LOGGER = logging.getLogger(__name__)

def print_prop_info(wpi, propName, value):
    #propInfo = next((x for x in wpi['properties'] if x['key'] == propName), None)
    propInfo = wpi_properties[propName]
    _LOGGER.debug(f"Property info: {propInfo}")
    propTitle = ""
    propDesc = ""
    propAlias = ""
    propRw = ""
    if 'rw' in propInfo:
        propRw = f", rw:{propInfo['rw']}"
    if 'alias' in propInfo:
        propAlias = f", alias:{propInfo['alias']}"
    if 'title' in propInfo:
        propTitle = propInfo['title']
    if 'description' in propInfo:
        propDesc = propInfo['description']
    print(f"- {propName} ({propInfo['jsonType']}{propRw}{propAlias}): {propTitle}")
    print(f"  Value: {value}")
    if propDesc:
        print(f"  Description: {propDesc}")
    if 'itemType' in propInfo:
        print(f"  Array item type: {propInfo['itemType']}")
    if 'min' in propInfo:
        print(f"  Minimum value: {propInfo['min']}")
    if 'max' in propInfo:
        print(f"  Maximum value: {propInfo['max']}")
    if 'example' in propInfo:
        print(f"  Example: {propInfo['example']}")

def watch_properties(name,value):
    if name in watching_properties:
        _LOGGER.info(f"Property {name} changed to {value}")

def watch_messages(wp,wsapp,msg,msg_json):
    if msg.type in watching_messages:
        _LOGGER.info(f"Message of type {msg.type} received: {msg}")

def cmd_get(wp, args):
    if len(args) != 1:
        _LOGGER.error(f"Wrong number of arguments: get <property>")
    elif not args[0] in wp.allProps:
        _LOGGER.error(f"Unknown property: {args[0]}")
    else:
        print(wp.allProps[args[0]])

def cmd_mqtt_connect(wp, args):
    if len(args) != 2:
        _LOGGER.error(f"Wrong number of arguments: mqtt-connect <host> <port>")
    else:
        host = args[0]
        port = int(args[1])
        mqtt_client.connect(host, port)

def cmd_mqtt_disconnect(wp):
    if mqtt_client.is_connected():
        _LOGGER.info(f"Disconnecting from MQTT server ...")
        mqtt_client.disconnect()

def cmd_set(wp, args):
    if len(args) != 2:
        _LOGGER.error(f"Wrong number of arguments: set <property> <value>")
    elif not args[0] in wp.allProps:
        _LOGGER.error(f"Unknown property: {args[0]}")
    else:
        if args[1].lower() in ["false","true"]:
            v=json.loads(args[1].lower())
        elif str(args[1]).isnumeric():
            v=int(args[1])
        elif str(args[1]).isdecimal():
            v=float(args[1])
        else:
            v=str(args[1])
        wp.send_update(args[0],v)

def cmd_watch_message(wp, name):
    if len(watching_messages) == 0:
        wp.register_message_callback(watch_messages)
    if name not in watching_messages:
        watching_messages.append(name)

def cmd_watch_property(wp, name):
    if len(watching_properties) == 0:
        wp.register_property_callback(watch_properties)
    if name not in watching_properties:
        watching_properties.append(name)

def cmd_watch(wp, args):
    if len(args) != 2:
        _LOGGER.error(f"Wrong number of arguments!")
    elif args[0] == "message" and not args[1] in wpi['messages']:
        _LOGGER.error(f"Unknown message type: {args[1]}")
    elif args[0] == "message":
        cmd_watch_message(wp, args[1])
    elif args[0] == "property" and not args[1] in wp.allProps:
        _LOGGER.error(f"Unknown property: {args[1]}")
    elif args[0] == "property":
        cmd_watch_property(wp, args[1])
    else:
        _LOGGER.error(f"Unknown watch type: {args[0]}")

def cmd_unwatch(wp,args):
    if len(args) != 2:
        _LOGGER.error(f"Wrong number of arguments!")
    elif args[0] == "message" and not args[1] in watching_messages:
        _LOGGER.warning(f"Message of type '{args[1]}' is not watched")
    elif args[0] == "message":
        watching_messages.remove(args[1])
        if len(watching_messages) == 0:
            wp.unregister_message_callback()
    elif args[0] == "property" and not args[1] in watching_properties:
        _LOGGER.warning(f"Property with name '{args[1]}' is not watched")
    elif args[0] == "property":
        watching_properties.remove(args[1])
        if len(watching_properties) == 0:
            wp.unregister_property_callback()
    else:
        _LOGGER.error(f"Unknown watch type: {args[0]}")

def process_command(wp, wpi, cmdline):
    """Process a Wattpilot shell command"""
    exit = False
    cmd_args = cmd_parser.parse_args(args=cmdline.strip().split(' '))
    cmd = cmd_args.cmd.strip()
    args = cmd_args.args
    if (cmd == ''):
        # Don't do anything
        exit = False
    elif (cmd == 'dump'):
        for propName, value in sorted(wp.allProps.items(), key=lambda x: x[0]): 
            print("{}: {}".format(propName, value))
    elif cmd == "exit":
        exit = True
    elif (cmd == 'get'):
        cmd_get(wp,args)
    elif (cmd == 'help'):
        print("Wattpilot Shell Commands:")
        print("  dump: Dump all property values")
        print("  exit: Exit the shell")
        print("  get <name>: Get a property value")
        print("  info: Print most important infos")
        print("  list [propsearch]: List all properties (starting with propsearch if given)")
        print("  mqtt-connect <host> <port>: Connect to MQTT server")
        print("  mqtt-disconnect: Disconnect from MQTT server")
        print("  set <name> <value>: Set a property value")
        print("  watch message <type>: Watch message of given message type")
        print("  watch property <name>: Watch value changes of given property name")
        print("  unwatch message <type>: Unwatch messages of given message type")
        print("  unwatch property <name>: Unwatch value changes of given property name")
    elif (cmd == 'info'):
        print(wp)
    elif (cmd == 'list'):
        print(f"List of available properties:")
        if len(args)==1:
            props = {k:v for k,v in wp.allProps.items() if k.startswith(args[0])}
        else:
            props = wp.allProps
        for propName, value in sorted(props.items()):
            print_prop_info(wpi,propName,value)
        print()
    elif (cmd == 'mqtt-connect'):
        cmd_mqtt_connect(wp, args)
    elif (cmd == 'mqtt-disconnect'):
        cmd_mqtt_disconnect(wp)
    elif (cmd == 'set'):
        cmd_set(wp, args)
    elif (cmd == 'watch'):
        cmd_watch(wp,args)
    elif (cmd == 'unwatch'):
        cmd_unwatch(wp,args)
    else:
        print(f"Unknown command: {cmd}")
    return exit

def complete(text,state):
    """Simple readline completer"""
    volcab = ['dump','exit','get','help','info','list','set']
    results = [x for x in volcab if x.startswith(text)] + [None]
    return results[state]

def wait_timeout(fn, timeout):
    """Generic timeout waiter"""
    t = 0
    within_timeout = True
    while not fn() and t<timeout:
        sleep(1)
        t += 1
    if t>=timeout:
        within_timeout = False
    return within_timeout

# MQTT functions

def mqtt_message(wp,wsapp,msg,msg_json):
    if mqtt_client == None:
        _LOGGER.debug(f"Skipping MQTT message publishing.")
        return
    if msg.type not in ["fullStatus", "deltaStatus"]:
        _LOGGER.debug(f"Skipping MQTT message publishing of type {msg.type}.")
        return
    msg_dict = json.loads(msg_json)
    if MQTT_PUBLISH_MESSAGES == "true":
        message_topic = MQTT_MESSAGE_TOPIC_PATTERN \
            .replace("{baseTopic}",MQTT_BASE_TOPIC) \
            .replace("{serialNumber}",wp.serial) \
            .replace("{messageType}",msg.type)
        mqtt_client.publish(message_topic, msg_json)
    if MQTT_PUBLISH_PROPERTIES == "true":
        for prop_name, value in msg_dict["status"].items():
            if MQTT_WATCH_PROPERTIES == [] or prop_name in MQTT_WATCH_PROPERTIES:
                _LOGGER.debug(f"Publishing property '{prop_name}' with value '{value}' to MQTT ...")
                property_topic = MQTT_PROPERTY_READ_TOPIC_PATTERN \
                    .replace("{baseTopic}",MQTT_BASE_TOPIC) \
                    .replace("{serialNumber}",wp.serial) \
                    .replace("{propName}",prop_name)
                mqtt_client.publish(property_topic, json.dumps(value))

# Timeout config:
WATTPILOT_CONNECT_TIMEOUT = int(os.environ.get('WATTPILOT_CONNECT_TIMEOUT','30'))
WATTPILOT_INITIALIZED_TIMEOUT = int(os.environ.get('WATTPILOT_INITIALIZED_TIMEOUT','30'))

# Globals:
watching_properties = []
watching_messages = []
mqtt_client = None

# MQTT config:
MQTT_ENABLED = os.environ.get('MQTT_ENABLED','false')
if MQTT_ENABLED == "true":
    MQTT_CLIENT_ID = os.environ.get('MQTT_CLIENT_ID','wattpilot2mqtt')
    MQTT_HA_DISCOVERY_ENABLED = os.environ.get('MQTT_HA_DISCOVERY_ENABLED','false')
    MQTT_HOST = os.environ.get('MQTT_HOST')
    MQTT_PORT = int(os.environ.get('MQTT_PORT','1883'))
    MQTT_BASE_TOPIC = os.environ.get('MQTT_BASE_TOPIC','wattpilot')
    MQTT_PUBLISH_MESSAGES = os.environ.get('MQTT_PUBLISH_MESSAGES','true')
    MQTT_MESSAGE_TOPIC_PATTERN = os.environ.get('MQTT_MESSAGE_TOPIC_PATTERN','{baseTopic}/{serialNumber}/messages/{messageType}')
    MQTT_PUBLISH_PROPERTIES = os.environ.get('MQTT_PUBLISH_PROPERTIES','false')
    MQTT_WATCH_PROPERTIES = os.environ.get('MQTT_WATCH_PROPERTIES','amp car fna lmo sse').split(sep=' ')
    MQTT_PROPERTY_READ_TOPIC_PATTERN = os.environ.get('MQTT_PROPERTY_READ_TOPIC_PATTERN','{baseTopic}/{serialNumber}/properties/{propName}')
    MQTT_PROPERTY_SET_TOPIC_PATTERN = os.environ.get('MQTT_PROPERTY_SET_TOPIC_PATTERN','{baseTopic}/{serialNumber}/properties/{propName}/set')
    mqtt_client = mqtt.Client(MQTT_CLIENT_ID)
    mqtt_client.connect(MQTT_HOST, MQTT_PORT)


# Set debug level:
logging.basicConfig(level=os.environ.get('WATTPILOT_DEBUG_LEVEL','INFO'))

# Setup readline
readline.parse_and_bind("tab: complete")
readline.set_completer(complete) 

# Commandline argument parser:
parser = argparse.ArgumentParser()
parser.add_argument("ip", help = "IP of Wattpilot Device", nargs="?")
parser.add_argument("password", help = "Password of Wattpilot", nargs="?")
parser.add_argument("cmdline", help = "Optional shell command", nargs="?")
args = parser.parse_args()

# Wattpilot shell command parser:
cmd_parser = argparse.ArgumentParser()
cmd_parser.add_argument("cmd", help = "Command")
cmd_parser.add_argument("args", help = "Arguments", nargs='*')

# Read Wattpilot config:
with open("wattpilot.yaml", 'r') as stream:
    try:
        wpi=yaml.safe_load(stream)
        wpi_messages = dict(zip(
            [x["key"] for x in wpi["messages"]],
            [x for x in wpi["messages"]],
        ))
        wpi_properties = dict(zip(
            [x["key"] for x in wpi["properties"]],
            [x for x in wpi["properties"]],
        ))
    except yaml.YAMLError as exc:
        print(exc)

# Connect to Wattpilot:
ip = args.ip or os.environ.get('WATTPILOT_HOST')
password = args.password or os.environ.get('WATTPILOT_PASSWORD')
wp = wattpilot.Wattpilot(ip,password)

# Enable MQTT integration:
if MQTT_ENABLED == "true":
    _LOGGER.debug(f"Registering message callback for MQTT integration.")
    wp.register_message_callback(mqtt_message)
wp.connect()

# Wait for connection and initializon:
wait_timeout(lambda: wp.connected, WATTPILOT_CONNECT_TIMEOUT) or exit("ERROR: Timeout while connecting to Wattpilot!")
wait_timeout(lambda: wp.allPropsInitialized, WATTPILOT_INITIALIZED_TIMEOUT) or exit("ERROR: Timeout while waiting for property initialization!")

# Process commands:
if args.cmdline:
    # Process commands passed on the commandline
    process_command(wp,wpi,args.cmdline)
else:
    # Start interactive shell
    process_command(wp,wpi,"info")
    exit=False
    while not exit:
        try:
            command = input('> ')
        except EOFError as e:
            command = "exit"
            print()
        exit = process_command(wp,wpi,command)
